### PR TITLE
Rmct phase 1

### PR DESCRIPTION
**Pull request by nsmohan**
Closing Issue #33 . 

Created a socket library where an socket object can be created as a server or client. 

--> Server (Serve data on network)
--> Client (Active listener; listen for data on network and report)

Created unittest to verify socket functionality.

### CODE REVIEW

File: lib/NMT_sock.cpp [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+     *  @return    NMT_result
+     */
+
+    NMT_log_write(DEBUG, (char *)"> ");
+    /* Initialize Varibles */
+    unsigned int opt = 1; 
+    NMT_result result = OK;
+
+    /* Create socket for sending/receiving datagrams */
+    if ((this->sock = socket(PF_INET, SOCK_DGRAM, IPPROTO_UDP)) < 0)
```
Comment (ChemistryMickey) [inline lib/NMT_sock.cpp]: Sorry, were do the arguments to socket get declared?
Are they compiler-defined macros?

File: lib/NMT_sock.cpp [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+     */
+
+    NMT_log_write(DEBUG, (char *)"> ");
+
+    /* Initialize Varibles */
+    NMT_result result = OK;
+    char msgbuf[MAX_BUFFER_SIZE];
+    socklen_t adder_len = sizeof(this->my_address);
+
+    int nbytes = recvfrom(this->sock, msgbuf, MAX_BUFFER_SIZE,
```
Comment (ChemistryMickey) [inline lib/NMT_sock.cpp]: Is there a risk to always pulling the MAX_BUFFER_SIZE?

File: lib/NMT_sock.cpp [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+    /* Exit the function */
+    NMT_log_write(DEBUG, (char *)"< result=%s", result_e2s[result]);
+    return result;
+}
+
+NMT_result NMT_sock_multicast::NMT_read_socket(char **message)
+{
+
+    /*!
+     *  @brief     Function to send message on socket
```
Comment (ChemistryMickey) [inline lib/NMT_sock.cpp]: "Function to read message from socket"
?

File: tst/unittest_NMT_sock.py [hunk 4]
```diff
@@ -48,18 +39,65 @@ def setUp(self):
...
+        print ("Message from second client socket=%s"%message_rx2)
+        self.assertEqual(message_tx, message_rx)
+        self.assertEqual(message_tx, message_rx2)
+
+    def test_02(self):
+
+        ##
+         # @test Test scenario where no message is recieved on the socket
+         # @step 1. Create 2 instnaces of the socket object (Server and Client)
+         # @step 2. CLient listens on the socket for 100ms and returns NOk as nothing is recieved
```
Comment (ChemistryMickey) [inline tst/unittest_NMT_sock.py]: Just a couple typos
"instances"
"Clients"

File: tst/unittest_NMT_sock.py [hunk 5]
```diff
@@ -48,18 +39,65 @@ def setUp(self):
         self.log_file = "%s/%s.log"%(NMT_log_test.LOG_DIR, FNAME)
 
     def test_01(self):
+
+        ##
+         # @test Test that client is able to recieve data that the server sends
+         # @step 1. Create 3 instnaces of the socket object (Server and 2 Clients)
```
Comment (ChemistryMickey) [inline tst/unittest_NMT_sock.py]: Typo: "instances"

File: lib/NMT_sock.cpp (hunk 1 above)
Comment (nsmohan) [reply lib/NMT_sock.cpp]: Socket is part of the #include <arpa/inet.h> library.

File: lib/NMT_sock.cpp (hunk 2 above)
Comment (nsmohan) [reply lib/NMT_sock.cpp]: MAX_BUFFER_SIZE defines that maximum amount of data were ready to receive.

File: lib/NMT_sock.cpp (hunk 3 above)
Comment (nsmohan) [reply lib/NMT_sock.cpp]: Updated.

File: tst/unittest_NMT_sock.py (hunk 4 above)
Comment (nsmohan) [reply tst/unittest_NMT_sock.py]: Updated.

File: tst/unittest_NMT_sock.py (hunk 5 above)
Comment (nsmohan) [reply tst/unittest_NMT_sock.py]: updated.

**Comment by nsmohan** [comment]
For the time being we are using python 2.7 for the unittests. At some point in the future we will update to the latest python.

**Pull request merged**